geode-core/src/main/java/org/apache/geode/internal/cache/GemFireCacheImpl.java

```diff
@@ -3929,22 +3938,7 @@ public void setGatewayConflictResolver(GatewayConflictResolver resolver) {
 
   @Override
   public List<CacheServer> getCacheServers() {
-    List<CacheServer> cacheServersWithoutReceiver = null;
-    if (!this.allCacheServers.isEmpty()) {
-      for (CacheServerImpl cacheServer : this.allCacheServers) {
-        // If CacheServer is a GatewayReceiver, don't return as part of CacheServers
-        if (!cacheServer.isGatewayReceiver()) {
-          if (cacheServersWithoutReceiver == null) {
-            cacheServersWithoutReceiver = new ArrayList<>();
-          }
-          cacheServersWithoutReceiver.add(cacheServer);
-        }
-      }
-    }
-    if (cacheServersWithoutReceiver == null) {
-      cacheServersWithoutReceiver = Collections.emptyList();
-    }
-    return cacheServersWithoutReceiver;
+    return this.cacheServersNotGatewayReceivers;
```

Comment:
The only difference that isn't account for here is that in the original code if the caller modifies the List the changes are not reflected into this class. Now that we return the live List from this class we could be susceptible to changes made by the caller. Maybe we should be keeping an immutable List to return here?

Reply:
Good point. I have now made it immutable. getCacheServers has lots of callers so we will see if any of them modified the list returned.